File: docs/js/badges.js
/* Badge evaluation - the JS twin of engine/badges.py (SPEC.md §3).
 * Rules are data (docs/data/badges.json); unknown kinds never fire. */
(function (global) {
  "use strict";

  function epoch(ts) {
    return Date.parse(ts) / 1000;
  }

  function moduleCompletionTs(state) {
    var out = {};
    for (var i = 0; i < state.completed.length; i++) {
      out[state.completed[i].module] = state.completed[i].ts;
    }
    return out;
  }

  function earnedTs(rule, state) {
    var kind = rule && rule.kind;
    var events = state.events, done = state.done;
    var i, ids;

    if (kind === "first_event") {
      for (i = 0; i < state.counted.length; i++) {
        if (state.counted[i].type === "pass") return state.counted[i].ts;
      }
      return null;
    }
    if (kind === "exercise") {
      return done[rule.id] ? done[rule.id].ts : null;
    }
    if (kind === "exercises_all") {
      ids = rule.ids || [];
      if (!ids.length) return null;
      var max = "";
      for (i = 0; i < ids.length; i++) {
        if (!done[ids[i]]) return null;
        if (done[ids[i]].ts > max) max = done[ids[i]].ts;
      }
      return max;
    }
    if (kind === "modules_all") {
      var byMod = moduleCompletionTs(state);
      ids = rule.ids || [];
      if (!ids.length) return null;
      var maxTs = "";
      for (i = 0; i < ids.length; i++) {
        if (!byMod[ids[i]]) return null;
        if (byMod[ids[i]] > maxTs) maxTs = byMod[ids[i]];
      }
      return maxTs;
    }
    if (kind === "devices") {
      var need = rule.count || 2, seen = {}, n = 0;
      for (i = 0; i < events.length; i++) {
        if (!seen[events[i].device]) { seen[events[i].device] = 1; n++; }
        if (n >= need) return events[i].ts;
      }
      return null;
    }
    if (kind === "gap_return") {
      var gap = (rule.days || 7) * 86400;
      var prev = null;
      for (i = 0; i < events.length; i++) {
        var t = epoch(events[i].ts);
        if (prev !== null && t - prev >= gap) return events[i].ts;
        prev = prev === null ? t : Math.max(prev, t);
      }
      return null;
    }
    return null; // unknown kind
  }

  function evaluate(defs, state) {
    var earned = [];
    for (var i = 0; i < defs.length; i++) {
      var ts = earnedTs(defs[i].rule || {}, state);
      if (ts !== null) {
        var row = {};
        for (var k in defs[i]) if (k !== "rule") row[k] = defs[i][k];
        row.earned_ts = ts;
        earned.push(row);
      }
    }
    earned.sort(function (a, b) {
      return a.earned_ts < b.earned_ts ? -1 : a.earned_ts > b.earned_ts ? 1 : 0;
    });
    return earned;
  }

  function nextTeasers(defs, state, limit) {
    limit = limit || 3;
    var have = {};
    var earned = evaluate(defs, state);
    for (var i = 0; i < earned.length; i++) have[earned[i].key] = 1;
    var out = [];
    for (i = 0; i < defs.length && out.length < limit; i++) {
      if (!have[defs[i].key]) {
        out.push({
          key: defs[i].key, name: defs[i].name, tier: defs[i].tier,
          icon: defs[i].icon, desc: defs[i].desc,
        });
      }
    }
    return out;
  }

  global.CodinBadges = { evaluate: evaluate, nextTeasers: nextTeasers };
})(typeof window !== "undefined" ? window : globalThis);

File: engine/badges.py
"""Badge evaluation. Mirrors SPEC.md §3; docs/js/badges.js is the twin.

Badges are derived, never stored: definitions live in
docs/data/badges.json, and adding a rule there awards it retroactively
from history. Unknown rule kinds are ignored (forward compatibility -
inventing a new kind is Workshop milestone M4 territory).
"""

import json
from pathlib import Path

from . import rules


def load_defs(repo_root):
    path = Path(repo_root) / "docs" / "data" / "badges.json"
    if not path.exists():
        return []
    return json.loads(path.read_text(encoding="utf-8"))


def _module_completion_ts(state):
    return {c["module"]: c["ts"] for c in state["completed"]}


def _earned_ts(rule, state):
    """-> the ts the rule fired at, or None if it hasn't. SPEC §3.1."""
    kind = rule.get("kind")
    events, done = state["events"], state["done"]

    if kind == "first_event":
        for ev in state["counted"]:
            if ev["type"] == "pass":
                return ev["ts"]
        return None

    if kind == "exercise":
        ev = done.get(rule.get("id"))
        return ev["ts"] if ev else None

    if kind == "exercises_all":
        ids = rule.get("ids", [])
        if ids and all(i in done for i in ids):
            return max(done[i]["ts"] for i in ids)
        return None

    if kind == "modules_all":
        by_mod = _module_completion_ts(state)
        ids = rule.get("ids", [])
        if ids and all(i in by_mod for i in ids):
            return max(by_mod[i] for i in ids)
        return None

    if kind == "devices":
        need = rule.get("count", 2)
        seen = set()
        for ev in events:
            seen.add(ev["device"])
            if len(seen) >= need:
                return ev["ts"]
        return None

    if kind == "gap_return":
        gap = rule.get("days", 7) * 86400
        prev = None
        for ev in events:
            t = _epoch(ev["ts"])
            if prev is not None and t - prev >= gap:
                return ev["ts"]
            prev = t if prev is None else max(prev, t)
        return None

    return None  # unknown kind: never fires


def _epoch(ts):
    """Seconds since epoch for a SPEC ts (UTC, seconds precision)."""
    from datetime import datetime, timezone

    return int(
        datetime.strptime(ts, "%Y-%m-%dT%H:%M:%SZ")
        .replace(tzinfo=timezone.utc)
        .timestamp()
    )


def evaluate(defs, state):
    """-> earned badges [{key,name,tier,icon,desc,earned_ts}] by earn time."""
    earned = []
    for b in defs:
        ts = _earned_ts(b.get("rule", {}), state)
        if ts is not None:
            row = dict(b)
            row.pop("rule", None)
            row["earned_ts"] = ts
            earned.append(row)
    earned.sort(key=lambda b: b["earned_ts"])
    return earned


def evaluate_all(repo_root, state):
    return evaluate(load_defs(repo_root), state)


def next_teasers(defs, state, limit=3):
    """Up to `limit` not-yet-earned badges worth showing, in file order.

    File order in badges.json is authored roughly by reachability, so
    the first unearned entries are the honest 'next achievable' set.
    """
    earned_keys = {b["key"] for b in evaluate(defs, state)}
    out = []
    for b in defs:
        if b["key"] in earned_keys:
            continue
        out.append({k: b[k] for k in ("key", "name", "tier", "icon", "desc") if k in b})
        if len(out) >= limit:
            break
    return out
